src/actions/exchangeActions.js

```diff
@@ -263,49 +270,84 @@ export const addExchangeAllowanceIfNeededAction = ({
       enabled: false,
     };
 
+    const senderAccount = findAccountByAddress(from, accountsSelector(getState()));
+    if (!senderAccount) {
+      reportErrorLog('addExchangeAllowanceIfNeededAction failed: no account found', { transaction });
+      return;
+    }
+
+    const senderAccountId = getAccountId(senderAccount);
+    const allowances = allAccountsAllowancesSelector(getState());
+
     // filter pending for current provider and asset match to override failed transactions
-    const allowances = _allowances.filter(({
-      provider: _provider,
-      fromAssetCode: _fromAssetCode,
-      toAssetCode: _toAssetCode,
-    }) => fromAssetCode !== _fromAssetCode && toAssetCode !== _toAssetCode && provider !== _provider);
+    const updatedAllowances = Object.keys(allowances).reduce((updated, allowancesAccountId) => {
+      if (senderAccountId !== allowancesAccountId) return updated;
 
-    allowances.push(allowance);
+      const accountAllowances = allowances[allowancesAccountId] || [];
+
+      const updatedAccountAllowances = accountAllowances
+        .filter(({
+          provider: _provider,
+          fromAssetCode: _fromAssetCode,
+          toAssetCode: _toAssetCode,
+        }) => fromAssetCode !== _fromAssetCode && toAssetCode !== _toAssetCode && provider !== _provider)
+        .concat(allowance);
+
+      return { ...updated, [allowancesAccountId]: updatedAccountAllowances };
+    }, {});
 
     dispatch({
-      type: ADD_EXCHANGE_ALLOWANCE,
-      payload: allowance,
+      type: ADD_ACCOUNT_EXCHANGE_ALLOWANCE,
+      payload: { allowance, accountId: senderAccountId },
     });
-    dispatch(saveDbAction('exchangeAllowances', { allowances }, true));
+
+    dispatch(saveDbAction('exchangeAllowances', { allowances: updatedAllowances }, true));
   };
 };
 
-export const enableExchangeAllowanceByHashAction = (transactionHash: string) => {
-  return async (dispatch: Dispatch, getState: GetState) => {
-    const { exchange: { data: { allowances: _allowances = [] } } } = getState();
-    const allowance = _allowances.find(
+export const enableExchangeAllowanceByHashAction = (transactionHash: string, senderAccountId: string) => {
+  return (dispatch: Dispatch, getState: GetState) => {
+    const allowances = allAccountsAllowancesSelector(getState());
+
+    const combinedAllowances = Object.keys(allowances).reduce((combined, accountId) => [
+      ...combined,
+      ...allowances[accountId],
+    ], []);
```

Comment:
This should be the same, but more idiomatic.
```suggestion
    const combinedAllowances = Object.keys(allowances).flatMap((accountId) => allowances[accountId]);
```